storage/grpc_client_test.go

```diff
@@ -218,3 +225,162 @@ func TestBytesCodecV2(t *testing.T) {
+			return writeStream, nil
+		})
+
+	c, err := NewGRPCClient(ctx, option.WithoutAuthentication(), option.WithGRPCDialOption(streamInterceptor))
+	if err != nil {
+		t.Fatalf("NewGRPCClient: %v", err)
+	}
+
+	w := gRPCWriter{
+		c:         (c.tc).(*grpcStorageClient),
+		buf:       []byte("012345689abcdefg"),
+		chunkSize: 10,
+		ctx:       ctx,
+		attrs:     &ObjectAttrs{},
+	}
+
+	maxAttempts := 4 // be careful changing this value; some tests depend on it
+
+	retriableErr := status.Errorf(codes.Internal, "a retriable error")
+	nonretriableErr := status.Errorf(codes.PermissionDenied, "a non-retriable error")
+	testCases := []struct {
+		desc            string
+		recvErrs        []error
+		sendErrs        []error
+		closeSend       bool // if true, will mimic finishing the upload so CloseSend will be called
+		expectedRetries int
+		expectedErr     error // final err from uploadBuffer should wrap this err
+	}{
+		{
+			desc:            "retriable error on receive",
+			recvErrs:        []error{retriableErr, retriableErr, retriableErr, retriableErr},
+			expectedRetries: maxAttempts - 1,
+			expectedErr:     retriableErr,
+		},
+		{
+			desc:            "retriable error on send",
+			sendErrs:        []error{io.EOF},
+			recvErrs:        []error{retriableErr, retriableErr, retriableErr, retriableErr},
+			expectedRetries: maxAttempts - 1,
+			expectedErr:     retriableErr,
+		},
+		{
+			desc:            "non-retriable error on send",
+			sendErrs:        []error{io.EOF},
+			recvErrs:        []error{nonretriableErr},
+			expectedRetries: 0,
+			expectedErr:     nonretriableErr,
+		},
+		{
+			desc:            "non-retriable err after send closed",
+			recvErrs:        []error{nil, nonretriableErr},
+			expectedRetries: 0,
+			expectedErr:     nonretriableErr,
+			closeSend:       true,
+		},
+		{
+			desc:            "retriable err after send closed",
+			recvErrs:        []error{nil, retriableErr, retriableErr, retriableErr, retriableErr},
+			expectedRetries: maxAttempts - 1,
+			expectedErr:     retriableErr,
+			closeSend:       true,
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.desc, func(t *testing.T) {
+			backoff := &mockBackoff{}
+			retryConfig := newUploadBufferRetryConfig(&settings{})
+			retryConfig.config.maxAttempts = &maxAttempts
+			retryConfig.config.backoff = backoff
+
+			writeStream.RecvMsgErrToReturn = test.recvErrs
+			writeStream.SendMsgErrToReturn = test.sendErrs
+
+			_, _, err = w.uploadBuffer(0, 0, test.closeSend, retryConfig)
+			if !errors.Is(err, test.expectedErr) {
+				t.Fatalf("uploadBuffer, want err to wrap: %v, got err: %v", test.expectedErr, err)
+			}
+
+			if got, want := backoff.pauseCallsCount, test.expectedRetries; got != want {
+				t.Errorf("backoff.Pause was called %d times, expected %d", got, want)
+			}
+		})
+	}
+}
+
+type MockWriteStream struct {
```

Comment:
This looks useful to test other behavior for writer; we should file a ticket to add.